Panic on linux during wait

A user on a linux machine reported that using `--wait` resulted in a go panic. Doing it without `--wait` worked just fine for them though.

```
$ ./carina-linux-amd64 create --wait testing123
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x405590]

goroutine 1 [running]:
main.(*CreateCommand).Create(0xc820070960, 0xc8200ae800, 0x7f1337f93000, 0xc82045f000)
	/Users/kyle6475/code/src/github.com/rackerlabs/carina/main.go:246 +0x3f0
main.(*CreateCommand).Create-fm(0xc8200ae800, 0x0, 0x0)
	/Users/kyle6475/code/src/github.com/rackerlabs/carina/main.go:116 +0x38
gopkg.in/alecthomas/kingpin%2ev2.(*actionMixin).applyActions(0xc82009e2d0, 0xc8200ae800, 0x0, 0x0)
	/Users/kyle6475/code/src/gopkg.in/alecthomas/kingpin.v2/actions.go:28 +0x7b
gopkg.in/alecthomas/kingpin%2ev2.(*Application).applyActions(0xc82007dba0, 0xc8200ae800, 0x0, 0x0)
	/Users/kyle6475/code/src/gopkg.in/alecthomas/kingpin.v2/app.go:496 +0x15b
gopkg.in/alecthomas/kingpin%2ev2.(*Application).execute(0xc82007dba0, 0xc8200ae800, 0x0, 0x0, 0x0, 0x0)
	/Users/kyle6475/code/src/gopkg.in/alecthomas/kingpin.v2/app.go:338 +0x250
gopkg.in/alecthomas/kingpin%2ev2.(*Application).Parse(0xc82007dba0, 0xc820074070, 0x5, 0x5, 0x0, 0x0, 0x0, 0x0)
	/Users/kyle6475/code/src/gopkg.in/alecthomas/kingpin.v2/app.go:139 +0x25e
main.main()
	/Users/kyle6475/code/src/github.com/rackerlabs/carina/main.go:325 +0x7d

goroutine 26 [IO wait]:
net.runtime_pollWait(0x7f1337f98808, 0x72, 0xc8200760a0)
	/usr/local/Cellar/go/1.5.1/libexec/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc8200c6ca0, 0x72, 0x0, 0x0)
	/usr/local/Cellar/go/1.5.1/libexec/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc8200c6ca0, 0x0, 0x0)
	/usr/local/Cellar/go/1.5.1/libexec/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc8200c6c40, 0xc82047e000, 0x400, 0x400, 0x0, 0x7f1337f93028, 0xc8200760a0)
	/usr/local/Cellar/go/1.5.1/libexec/src/net/fd_unix.go:232 +0x23a
net.(*conn).Read(0xc82008c1e8, 0xc82047e000, 0x400, 0x400, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.5.1/libexec/src/net/net.go:172 +0xe4
crypto/tls.(*block).readFromUntil(0xc8200d7170, 0x7f1337f98b78, 0xc82008c1e8, 0x5, 0x0, 0x0)
	/usr/local/Cellar/go/1.5.1/libexec/src/crypto/tls/conn.go:455 +0xcc
crypto/tls.(*Conn).readRecord(0xc8200a9340, 0x927817, 0x0, 0x0)
	/usr/local/Cellar/go/1.5.1/libexec/src/crypto/tls/conn.go:540 +0x2d1
crypto/tls.(*Conn).Read(0xc8200a9340, 0xc820481000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.5.1/libexec/src/crypto/tls/conn.go:901 +0x167
net/http.noteEOFReader.Read(0x7f1337fdbb68, 0xc8200a9340, 0xc8203cacb8, 0xc820481000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.5.1/libexec/src/net/http/transport.go:1370 +0x67
net/http.(*noteEOFReader).Read(0xc8204a4020, 0xc820481000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	<autogenerated>:126 +0xd0
bufio.(*Reader).fill(0xc820075440)
	/usr/local/Cellar/go/1.5.1/libexec/src/bufio/bufio.go:97 +0x1e9
bufio.(*Reader).Peek(0xc820075440, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.5.1/libexec/src/bufio/bufio.go:132 +0xcc
net/http.(*persistConn).readLoop(0xc8203cac60)
	/usr/local/Cellar/go/1.5.1/libexec/src/net/http/transport.go:876 +0xf7
created by net/http.(*Transport).dialConn
	/usr/local/Cellar/go/1.5.1/libexec/src/net/http/transport.go:685 +0xc78

goroutine 27 [select]:
net/http.(*persistConn).writeLoop(0xc8203cac60)
	/usr/local/Cellar/go/1.5.1/libexec/src/net/http/transport.go:1009 +0x40c
created by net/http.(*Transport).dialConn
	/usr/local/Cellar/go/1.5.1/libexec/src/net/http/transport.go:686 +0xc9d
```

On a side note, I should probably build the release binaries in a clean docker environment to not leak out my local dev setup. :wink:

--- a/main.go
+++ b/main.go
@@ -365,6 +365,8 @@ func (carina *Command) informLatest(pc *kingpin.ParseContext) error {
 	return nil
 }
 
+const httpTimeout = time.Second * 10
+
 // Auth does the authentication
 func (carina *Command) Auth(pc *kingpin.ParseContext) (err error) {
 
@@ -384,8 +386,12 @@ func (carina *Command) Auth(pc *kingpin.ParseContext) (err error) {
 		}
 	}
 
+	// Short circuit if the cache is not enabled
 	if !carina.CacheEnabled {
 		carina.ClusterClient, err = libcarina.NewClusterClient(carina.Endpoint, carina.Username, carina.APIKey)
+		if err != nil {
+			carina.ClusterClient.Client.Timeout = httpTimeout
+		}
 		return err
 	}
 
@@ -393,7 +399,7 @@ func (carina *Command) Auth(pc *kingpin.ParseContext) (err error) {
 
 	if ok {
 		carina.ClusterClient = &libcarina.ClusterClient{
-			Client:   &http.Client{},
+			Client:   &http.Client{Timeout: httpTimeout},
 			Username: carina.Username,
 			Token:    token,
 			Endpoint: carina.Endpoint,

--- a/main.go
+++ b/main.go
@@ -365,7 +365,7 @@ func (carina *Command) informLatest(pc *kingpin.ParseContext) error {
 	return nil
 }
 
-const httpTimeout = time.Second * 10
+const httpTimeout = time.Second * 45
 
 // Auth does the authentication
 func (carina *Command) Auth(pc *kingpin.ParseContext) (err error) {

--- a/main.go
+++ b/main.go
@@ -515,17 +515,30 @@ const StatusRebuildingSwarm = "rebuilding-swarm"
 // Does an func against a cluster then returns the new cluster representation
 func (carina *WaitClusterCommand) clusterApplyWait(op clusterOp) (err error) {
 	cluster, err := op(carina.ClusterName)
+	if err != nil {
+		return err
+	}
 
 	if carina.Wait {
+		var status string
+
+		status = cluster.Status
+
 		time.Sleep(startupFudgeFactor)
 		// Transitions past point of "new" or "building" are assumed to be states we
 		// can stop on.
-		for cluster.Status == StatusNew || cluster.Status == StatusBuilding || cluster.Status == StatusRebuildingSwarm {
+		for status == StatusNew || status == StatusBuilding || status == StatusRebuildingSwarm {
 			time.Sleep(waitBetween)
 			cluster, err = carina.ClusterClient.Get(carina.ClusterName)
-			if err != nil {
-				break
+			if err != nil || cluster == nil {
+				// Assume we should reauth
+				carina.ClusterClient, err = libcarina.NewClusterClient(carina.Endpoint, carina.Username, carina.APIKey)
+				if err != nil {
+					break
+				}
+				continue
 			}
+			status = cluster.Status
 		}
 	}
 
@@ -556,7 +569,8 @@ func (carina *CreateCommand) Create(pc *kingpin.ParseContext) (err error) {
 			Nodes:       nodes,
 			AutoScale:   carina.AutoScale,
 		}
-		return carina.ClusterClient.Create(c)
+		cluster, err := carina.ClusterClient.Create(c)
+		return cluster, err
 	})
 }
 

--- a/main.go
+++ b/main.go
@@ -365,7 +365,7 @@ func (carina *Command) informLatest(pc *kingpin.ParseContext) error {
 	return nil
 }
 
-const httpTimeout = time.Second * 45
+const httpTimeout = time.Second * 15
 
 // Auth does the authentication
 func (carina *Command) Auth(pc *kingpin.ParseContext) (err error) {
@@ -520,19 +520,25 @@ func (carina *WaitClusterCommand) clusterApplyWait(op clusterOp) (err error) {
 	}
 
 	if carina.Wait {
-		var status string
-
-		status = cluster.Status
-
 		time.Sleep(startupFudgeFactor)
+
+		carina.ClusterClient.Client = &http.Client{Timeout: httpTimeout}
+		cluster, err = carina.ClusterClient.Get(carina.ClusterName)
+		if err != nil {
+			return err
+		}
+
+		status := cluster.Status
+
 		// Transitions past point of "new" or "building" are assumed to be states we
 		// can stop on.
 		for status == StatusNew || status == StatusBuilding || status == StatusRebuildingSwarm {
 			time.Sleep(waitBetween)
+			// Assume go has held this connection live long enough
+			carina.ClusterClient.Client = &http.Client{Timeout: httpTimeout}
 			cluster, err = carina.ClusterClient.Get(carina.ClusterName)
 			if err != nil || cluster == nil {
 				// Assume we should reauth
-				carina.ClusterClient, err = libcarina.NewClusterClient(carina.Endpoint, carina.Username, carina.APIKey)
 				if err != nil {
 					break
 				}

